### ISSUE TITLE
Type inference seems to ignore reachability analysis

### ISSUE THREAD

**Issue by tapetersen** [microsoft/pyright#6850]
**Describe the bug**
Not important as such in any specific case but wanted to report.

While pyright detects that a section of code is unreachable it still uses these statements for type inference of a variable.

**Code or Screenshots**
In the following code pyright excludes the possibility of `a=5` since it detects the overwrite later but it has also marked that code as unreachable 

![image](https://github.com/microsoft/pyright/assets/380497/b56ec2f2-9c63-4ef4-8af2-9bd51abee37e)

```python
# pyright: strict

from typing import Any, Literal
from typing_extensions import assert_type, Never


def no_return() -> Never:
    raise Exception()

class CtxMgr:
    def __enter__(self):
        return self

    def __exit__(self, *exc_info: Any) -> bool:
        return True

a = None
with CtxMgr():
    a = 5
    no_return()
    a = 10

assert a is not None
assert_type(a, Literal[5])
```

[playground link ](https://pyright-play.net/?strict=true&locale=en&code=MQAgDgngTglg5gCwC4C4QGcmwMZILABQhAZlAPYC2ISEYMAdnCDBWGVEiAIL0QA0IADIwkAUygBDADYlyVGnUYB9UQA8x9dDDKbmrdpwnp04pEoWiBAOVEA3cYUcEAJqOIh6ZJVFFIArlD0ABQAlCAAtAB8IDb2UCiEIEkgkjAmIACiqtiiYEjawSFO2FJG6CAAwkiqALJw8YnJru5KKvRiUK1BJlLEIQkEyUMpvgH0GKK9TkPNIK1qIl09xAIAVGrYSgzEZGg8EGFRIABGZGRSA8PJPv6BIAAqUH6iThIgALwxOi8EAO4iCEq1TqUFCl2Sb0%2BAFZGklPN5RoFQrCQJCQABGAAMr2MplRzHKnk4Vm%2BhDKpnMtFEQQkAmEHWkAG0oQBdEJAA)

**VS Code extension or command-line**
Pylance language server 2023.12.1 (pyright a78e2b10)

**Comment by erictraut** [comment]
Good catch. Yes, there was a subtle (and longstanding) bug in the code flow analysis logic. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.344, which I just published. It will also be included in a future release of pylance.
